presto-main/src/main/java/com/facebook/presto/operator/RemoteProjectOperator.java

```diff
@@ -0,0 +1,196 @@
+        this.result = new CompletableFuture[projections.size()];
+    }
+
+    @Override
+    public OperatorContext getOperatorContext()
+    {
+        return operatorContext;
+    }
+
+    @Override
+    public boolean needsInput()
+    {
+        return !finishing && !processingPage();
+    }
+
+    @Override
+    public void addInput(Page page)
+    {
+        checkState(!finishing, "Operator is already finishing");
+        checkState(!processingPage(), "Still processing previous input");
+        requireNonNull(page, "page is null");
+        for (int channel = 0; channel < projections.size(); channel++) {
+            RowExpression projection = projections.get(channel);
+            Type type = projection.getType();
+            if (projection instanceof InputReferenceExpression) {
+                result[channel] = completedFuture(page.getBlock(((InputReferenceExpression) projection).getField()));
+            }
+            else if (projection instanceof CallExpression) {
+                CallExpression remoteCall = (CallExpression) projection;
+                result[channel] = functionManager.executeFunction(
+                        remoteCall.getFunctionHandle(),
+                        page,
+                        remoteCall.getArguments().stream()
+                                .map(InputReferenceExpression.class::cast)
+                                .map(InputReferenceExpression::getField)
+                                .collect(toImmutableList()));
+            }
+            else {
+                checkState(projection instanceof ConstantExpression, format("Does not expect expression type %s", projection.getClass()));
+            }
+        }
+    }
+
+    @Override
+    public Page getOutput()
+    {
+        if (resultReady()) {
+            Block[] blocks = new Block[result.length];
+            Page output;
+            try {
+                for (int i = 0; i < blocks.length; i++) {
+                    blocks[i] = result[i].get();
+                }
+                output = new Page(blocks);
+                Arrays.fill(result, null);
+                return output;
+            }
+            catch (InterruptedException ie) {
+                currentThread().interrupt();
+            }
+            catch (ExecutionException e) {
+                Throwable cause = e.getCause();
+                if (cause != null) {
+                    throwIfUnchecked(cause);
+                    throw new RuntimeException(cause);
+                }
+                throw new RuntimeException(e);
+            }
+        }
+        return null;
+    }
+
+    @Override
+    public void finish()
+    {
+        finishing = true;
+    }
+
+    @Override
+    public boolean isFinished()
+    {
+        return finishing && !processingPage();
+    }
+
+    private boolean processingPage()
+    {
+        for (int i = 0; i < result.length; i++) {
+            if (result[i] != null) {
```

Comment:
If I understand correctly, if `result` array contains non-null element, it means `addInput` was called after last `getOutput`: this is because in `getOutput`, `result` array will be filled with `null` (line 110).

If that's the case, does it make sense to just have a `AtomicBoolean` for this flag, say something like `inputsAdded` ? For two reason:
1. More efficient (probably not really important)
2. I was originally looking into the difference between `resultReady`, and wondering why we don't need to check `result[i].isDone()`

Reply:
I didn't quite get the second reason. I thought about the first one. Didn't do it cause feels performance is not enough reason to add an additional variable here. Maintaining two states separately has "maintenance cost" as well. I can add it if you prefer. Don't have strong opinion here.

Reply:
```
// result array will be filled with all null values after getOutput() get called
```